### PR TITLE
Feat/nodes with document handle

### PR DESCRIPTION
**Pull request by jaytaph**
I think this current setup sounds like a reasonable "stable" state for now which fixes some problems concerning document handling through the parser.

It does not fix anything in the parser-tests, since these changes are needed for "template" elements which aren't tested yet.

From this point, we can fix the parser adjust_insert_location() function which IS needed to make the template elements (among others), so it will pass more test.

### CODE REVIEW

File: src/html5_parser/parser/adoption_agency.rs [hunk 1]
```diff
@@ -12,13 +12,15 @@ pub enum AdoptionResult {
...
+impl<'stream> Html5Parser<'stream> {
+    /**
+     * When we talk about nodes, there are 3 contexts to consider:
+     *
+     * - The actual node data. This is called "node" in the code.
+     * - The node id. This is called "node_id" in the code.
+     * - The node index. This is called "node_idx" in the code. This is the index of the node in
+     *   either the open_elements or active_formatting_elements stack.
+     */
     pub fn run_adoption_agency(&mut self, token: &Token) -> AdoptionResult {
```
Comment (emwalker) [inline src/html5_parser/parser/adoption_agency.rs:24]: We should probably do a second pass soon and reintroduce some of the good changes, like this one, that are being reverted here (not necessarily in this PR).

File: src/html5_parser/parser/document.rs [hunk 2]
```diff
@@ -50,6 +85,29 @@ impl Document {
         }
     }
 
+    pub fn shared() -> DocumentHandle {
+        DocumentHandle(Rc::new(RefCell::new(Self::new())))
+    }
```
Comment (emwalker) [inline src/html5_parser/parser/document.rs:90]: The name `Document::shared()` is a little weird.  But the interface can be dialed in in a later PR.

**Review by emwalker — APPROVED** [review]
Most of the credit goes to @Kiyoshika for figuring out how to incorporate the ref-counted pointer into the existing code (most of this work is his).

I'm hoping the interface can be dialed in and revisited as we gain more experience using these smart pointers.  And it would be nice to reintroduce some of the changes that have been rolled back, such as the dropping of the macros and the touching up of the comments.  But this PR is a net improvement, and we can follow up with those things in later PRs.

**Comment by emwalker** [comment]
@jaytaph I suggest configuring the merge button to only allow squash and merge, so that we can preserve a linear commit history of atomic commits on `main`.  That will make using `git bisect` much easier in the future.

**Comment by jaytaph** [comment]
@emwalker my opinion about squash and merge is that it foes a rewrite which should be the author's responsibility. I've already rebased my tmp commits in this pr, and i think we could rebase again into 1 commit again. However, this is not always the case: there are many valid reasons why we want to have multiple commits inside a single pr (wspecially when it comes to bisecting), so i would suggest to do merges, but manually check if the commits actually make sense (this will work better at a later stage, as there are a lot of changes done presently.

**Comment by jaytaph** [comment]
Looking at it from mobile, i would rebase it to 3 commits in this case

File: src/html5_parser/parser/adoption_agency.rs (hunk 1 above)
Comment (jaytaph) [reply src/html5_parser/parser/adoption_agency.rs:24]: Agreed. I think that some of the reverts are due to a lot of manual fixups during rebasing of this branch on main, as there were a lot of changes done in both branches. This should (hopefully) not happen very often as codebase matures

**Comment by emwalker** [comment]
I think having authors rebase is sensible if we can maintain discipline.  As king commits are atomic and the tests pass, it will be straightforward to do a `git bisect` when problems come up.  If commits are merged that are just broken checkpoint commits, a `git bisect` will be laborious and challenging.

Some of this can be mitigated by keeping PRs small and focused (when possible).

**Comment by jaytaph** [comment]
@Kiyoshika any input?  I wouldn't mind to change it to squash/merge, and i won't be standing in the way if we need to do this, but i'm not directly for it :-)

**Comment by Kiyoshika** [comment]
> @Kiyoshika any input? I wouldn't mind to change it to squash/merge, and i won't be standing in the way if we need to do this, but i'm not directly for it :-)

I don't have a strong opinion on squash and merge, this is actually my first ever open source project I've contributed to so I have only dealt with my own simple commits in my own projects; I'm not aware of the benefits/drawbacks

**Comment by Kiyoshika** [comment]
Also just noticed that I forgot to sign one of my commits, I really need to update my default

**Review by Kiyoshika — APPROVED** [review]

**Comment by jaytaph** [comment]
I'll just squash commit this for now

**Pull request merged**